Upgrade Message Schema (Data) in Background

gasi-signal opened the pull request:
In order to avoid incurring long startup times, we migrate message schema (data) in the background using `window.requestIdleCallback` API. The migration moves attachment data from IndexedDB to disk and reduces load on our database which may cause data loss (#1589).

On my development profile, this migration reduced the IndexedDB directory from 33.4MB to 4.7MB.

gasi-signal commented on js/modules/types/attachment.js:
```diff
@@ -29,19 +29,17 @@ const { migrateDataToFileSystem } = require('./attachment/migrate_data_to_file_s
-  const hasValidFileName =
-    isString(rawAttachment.fileName) || rawAttachment.fileName === null;
-  return hasValidContentType && hasValidFileName;
+  return true;
```
Temporarily removed this check as you found `fileName === undefined`, @scottnonnenberg-signal.

scottnonnenberg-signal commented on preload.js:
```diff
@@ -113,8 +113,12 @@
   const readAttachmentData = Attachments.readData(attachmentsPath);
   const writeAttachmentData = Attachments.writeData(attachmentsPath);
 
+  const { IdleDetector} = require('./js/modules/idle_detector');
```
This is pretty awkward. I can see why we might want named exports, like for future-proofing as we add more things to files. But perhaps a better approach might be to choose a parent namespace of some kind, and put it inside that. Like `window.Signal.OS` does.

scottnonnenberg-signal commented on js/modules/idle_detector.js:
```diff
@@ -0,0 +1,46 @@
+
+  _scheduleNextCallback() {
+    this.stop();
+    this.handle = requestIdleCallback((deadline) => {
```
Might be nice to put a `window.` here to make it clear where this function is coming from.

gasi-signal commented:
@scottnonnenberg-signal Addressed feedback and rebased. Changes: https://github.com/signalapp/Signal-Desktop/pull/2162/files/85e478c2586fdbb60d32045aebe7995c7d88030c..e4d30ef6e2c307e1849c556efa2af4595b67ddc0

gasi-signal commented:
Rebased (no changes) and will merge before CI. Will address any unlikely CI issues after all three migration PRs are merged into `development`.

ByteHamster commented:
How long is the schema upgrade supposed to take? 24 seconds per message is wrong, I guess?

```
INFO  2018-04-21T12:42:24.478Z Upgrade message schema: 
{  
   "batchTotalDuration":24,
   "done":false,
   "fetchDuration":1,
   "lastProcessedIndex":"12ae1d98-bd16-72fd-8a0a-fb2b5288ca08",
   "newLastProcessedIndex":"12ae91be-1887-cee7-2116-42c7d2030420",
   "numMessagesProcessed":1,
   "saveDuration":21,
   "targetSchemaVersion":3,
   "upgradeDuration":1
}
```

Does not look like my drive is too busy:
```
> iostat -dx
Linux 4.15.15-1-ARCH (hostname)     04/21/18        _x86_64_        (8 CPU)

Device            r/s     w/s     rkB/s     wkB/s   rrqm/s   wrqm/s  %rrqm  %wrqm r_await w_await aqu-sz rareq-sz wareq-sz  svctm  %util
sdb            111.12   17.00   3480.28   1905.15     5.99    58.76   5.11  77.56    3.80  373.16   6.77    31.32   112.06   1.32  16.86
sda              9.70    7.93    211.93    282.95     0.21     9.33   2.08  54.07    0.63   26.73   0.22    21.84    35.70   0.88   1.56
```

scottnonnenberg-signal commented:
That's 24ms. Anyway, we're doing it very slowly in the background, whenever Chromium tells us we have some idle time to use. You shouldn't be worried about the speed.

ByteHamster commented:
Okay, thanks. I thought it's 24s because I left it running for 6 hours and it's still not finished. Those messages appear in the log every ~20 seconds.

I was hoping that it fixes the slow startup time. So I will just leave it running a little longer. Thanks for the reply
